compile.go

```diff
@@ -704,30 +700,74 @@ func ToQueryTime(value values.Value) (Time, error) {
 }
 
 type importer struct {
+	r    *runtime
 	pkgs map[string]*interpreter.Package
 }
 
 func (imp *importer) Copy() *importer {
-	packages := make(map[string]*interpreter.Package, len(imp.pkgs))
-	for k, v := range imp.pkgs {
-		packages[k] = v.Copy()
-	}
-	return &importer{
-		pkgs: packages,
+	return &importer{r: imp.r}
+}
+
+func (imp *importer) Import(path string) (semantic.MonoType, error) {
+	p, err := imp.ImportPackageObject(path)
+	if err != nil {
+		return semantic.MonoType{}, err
 	}
+	return p.Type(), nil
 }
 
-func (imp *importer) Import(path string) (semantic.MonoType, bool) {
-	p, ok := imp.pkgs[path]
+func (imp *importer) ImportPackageObject(path string) (*interpreter.Package, error) {
+	// If this package has been imported previously, return the import now.
+	if p, ok := imp.pkgs[path]; ok {
+		if p == nil {
+			return nil, errors.Newf(codes.Invalid, "detected cyclical import for package path %q", path)
+		}
+		return p, nil
+	}
+
+	// Mark down that we are currently evaluating this package
+	// so that we can detect a circular import.
+	if imp.pkgs == nil {
+		imp.pkgs = make(map[string]*interpreter.Package)
+	}
+	imp.pkgs[path] = nil
+
+	// If this package is part of the prelude, fill in a fake
+	// empty package to resolve cyclical imports.
+	for _, ppath := range prelude {
+		if ppath == path {
+			imp.pkgs[path] = interpreter.NewPackage(path)
+			break
+		}
+	}
+
+	// Find the package for the given import path.
+	semPkg, ok := imp.r.pkgs[path]
 	if !ok {
-		return semantic.MonoType{}, false
+		return nil, errors.Newf(codes.Invalid, "invalid import path %s", path)
+	}
+
+	// Construct the prelude scope from the prelude paths.
+	// If we are importing part of the prelude, we do not
+	// include it as part of the prelude and will stop
+	// including values as soon as we hit the prelude.
+	// This allows us to import all previous paths when loading
+	// the prelude, but avoid a circular import.
+	scope, err := imp.r.newScopeFor(path, imp)
```

Comment:
Does this mean we are re-evaluating the prelude with every import statement?

Reply:
The prelude only gets evaluated once and then cached for that execution. So it would be evaluated once per script, but not once per import.